### PR TITLE
feat: add Store API multi-currency support

### PR DESCRIPTION
**Pull request by frosso**
Fixes #

#### Changes proposed in this Pull Request

Adding multi-currency support to the Store API.
Tangentially, this also helps fix the multi-currency feature with the tokenized cart.

#### Testing instructions

The main purpose of this is to support the multi-currency feature with the Store API.
You can check this behavior by going into a block-based page (e.g.: block-based checkout page) and making this example request on a store with USD as the main currency and with EUR enabled:
```js
window.wp.apiFetch( {
    method: 'GET',
    path: '/wc/store/v1/cart?currency=EUR',
    // omitting credentials, to create a new cart object separate from the user's cart.
    credentials: 'omit',
} )
```

You'll notice that in the response, the `totals.currency_code` string corresponds to the EUR currency.

To further test this behavior with the tokenized cart with PRBs:
- Enable the "Enable Cart-Token implementation for PRBs" flag in your dev tools (you'll need to update them)
- Ensure you have Google Pay/Apple Pay enabled in the merchant's settings
- Ensure you have the multi-currency feature enabled, with a few currencies enabled (e.g.: USD & EUR)
- Ensure the selected currency is not the store's "default" currency (i.e.: as a shopper, select EUR if your store's default currency is USD)
- Go to a product page on your site
- Increase the product's quantity (so that the cached data from the backend is ignored)
- Click on the PRB button, so that the dialog opens
- The price of the product should match the price displayed on the PRB dialog

With this product (the single unit price is 14EUR/15USD):
Screenshot 2024-05-14 at 9 35 18 PM

Before - selected currency is EUR, but displayed currency in the PRB dialog is USD & price is 2x USD single unit price:
Untitled

After - selected currency is EUR, displayed currency in the PRB dialog is accurate and price is accurate:
Screenshot 2024-05-14 at 9 38 31 PM

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8816` or branch name `feat/add-multi-currency-support-to-store-api` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=feat/add-multi-currency-support-to-store-api) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 3149c01386a79141b9b280f2d89e732839f2ef55
- Build time: 2024-05-17 09:24:21 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/class-wc-payments-utils.php [hunk 1]
```diff
@@ -1073,6 +1092,30 @@ public static function is_cart_block(): bool {
 		return has_block( 'woocommerce/cart' ) || ( wp_is_block_theme() && is_cart() );
 	}
 
+	/**
+	 * Returns true if the request that's currently being processed is a Store API request, false
+	 * otherwise.
+	 *
+	 * @return bool True if request is a Store API request, false otherwise.
+	 */
+	public static function is_store_api_request(): bool {
```
Comment (frosso) [inline includes/class-wc-payments-utils.php:1101]: Moved from `includes/woopay/class-woopay-session.php`

File: includes/class-wc-payments-utils.php [hunk 2]
```diff
@@ -33,6 +33,25 @@ class WC_Payments_Utils {
 	 */
 	const FORCE_DISCONNECTED_FLAG_NAME = 'wcpaydev_force_disconnected';
 
+	/**
+	 * The Store API route patterns that should be handled by the WooPay session handler.
+	 */
+	const STORE_API_ROUTE_PATTERNS = [
```
Comment (frosso) [inline includes/class-wc-payments-utils.php:39]: Moved from `includes/woopay/class-woopay-session.php`

File: includes/woopay/class-woopay-session.php [hunk 3]
```diff
@@ -648,30 +629,6 @@ private static function get_woopay_verified_email_address() {
 		return $has_woopay_verified_email_address ? sanitize_text_field( wp_unslash( $_SERVER['HTTP_X_WOOPAY_VERIFIED_EMAIL_ADDRESS'] ) ) : null;
 	}
 
-	/**
-	 * Returns true if the request that's currently being processed is a Store API request, false
-	 * otherwise.
-	 *
-	 * @return bool True if request is a Store API request, false otherwise.
-	 */
-	private static function is_store_api_request(): bool {
```
Comment (frosso) [inline includes/woopay/class-woopay-session.php:657]: Moved to `includes/class-wc-payments-utils.php` for reusability

File: includes/woopay/class-woopay-session.php [hunk 4]
```diff
@@ -33,25 +33,6 @@ class WooPay_Session {
 
 	const STORE_API_NAMESPACE_PATTERN = '@^wc/store(/v[\d]+)?$@';
 
-	/**
-	 * The Store API route patterns that should be handled by the WooPay session handler.
-	 */
-	const STORE_API_ROUTE_PATTERNS = [
```
Comment (frosso) [inline includes/woopay/class-woopay-session.php:39]: Moved to `includes/class-wc-payments-utils.php` for reusability

File: includes/multi-currency/MultiCurrency.php [hunk 5]
```diff
@@ -244,11 +244,11 @@ public function init_hooks() {
 
 		$is_frontend_request = ! is_admin() && ! defined( 'DOING_CRON' ) && ! WC()->is_rest_api_request();
 
-		if ( $is_frontend_request ) {
+		if ( $is_frontend_request || \WC_Payments_Utils::is_store_api_request() ) {
```
Comment (frosso) [inline includes/multi-currency/MultiCurrency.php:247]: Ensures that the `currency` query string parameter can be used by the Store API

File: includes/multi-currency/Utils.php [hunk 6]
```diff
@@ -57,7 +57,7 @@ public function is_page_with_vars( array $pages, array $vars ): bool {
 	 * @return boolean
 	 */
 	public static function is_admin_api_request(): bool {
-		return 0 === stripos( wp_get_referer(), admin_url() ) && WC()->is_rest_api_request();
+		return 0 === stripos( wp_get_referer(), admin_url() ) && WC()->is_rest_api_request() && ! \WC_Payments_Utils::is_store_api_request();
```
Comment (frosso) [inline includes/multi-currency/Utils.php:60]: WC API requests were marked as "admin API" requests (which they aren't)

**Review by gpressutto5 — COMMENTED** [review]
As you can see in the images below, I encountered some problems, but I'm not sure if they happen because of this PR or because of the Cart-Token implementation for PRBs.

| | Apple Pay USD | Apple Pay EUR | Google Pay USD | Google Pay EUR |
| ----- | ----- | ----- | ----- | ----- |
| `develop` | image | image | image | image |
| This PR | image | image | image | image |



1. When I use Apple Pay, the amount is always Pending, even after loading the shipping fee.
2. Taxes are not being calculated.
3. The shipping fee is not itemized.

**Comment by frosso** [comment]
@gpressutto5 thank you for the report! The itemized amounts not being displayed is something I still need to fix on the main PR, unfortunately :( 
This PR should just fix the conversion of the amounts from one currency to the other - so that the currency selected by the customer (and the item amount) is the same as the currency displayed in the PRB dialog

**Review by gpressutto5 — APPROVED** [review]
The cart currency is switched as expected.

**Pull request merged**